### PR TITLE
Automatically add valid and non-existing repository when executed from command line

### PR DESCRIPTION
**Pull request by ggajos**
There are still open topics in #4259 but I believe this is fixing the original issue requested by @gingerbeardman. This change doesn't prompt user to add repo if this is a valid repository.

### CODE REVIEW

**Comment by vanfumin08** [comment]
Master

**Comment by vanfumin08** [comment]
Make software delivery more collaborative, reliable, and well-documented

File: app/src/lib/dispatcher/dispatcher.ts [hunk 1]
```diff
@@ -54,6 +54,7 @@ import { FetchType } from '../../lib/stores'
 import { PullRequest } from '../../models/pull-request'
 import { IAuthor } from '../../models/author'
 import { ITrailer } from '../git/interpret-trailers'
+import { isGitRepository } from '../git'
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: -

File: app/src/lib/dispatcher/dispatcher.ts [hunk 2]
```diff
@@ -54,6 +54,7 @@ import { FetchType } from '../../lib/stores'
 import { PullRequest } from '../../models/pull-request'
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: /

File: app/src/lib/dispatcher/dispatcher.ts [hunk 3]
```diff
@@ -54,6 +54,7 @@ import { FetchType } from '../../lib/stores'
...
 import { IAuthor } from '../../models/author'
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: n

File: app/src/lib/dispatcher/dispatcher.ts [hunk 4]
```diff
@@ -54,6 +54,7 @@ import { FetchType } from '../../lib/stores'
...
 import { ITrailer } from '../git/interpret-trailers'
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: '

File: app/src/lib/dispatcher/dispatcher.ts [hunk 5]
```diff
@@ -848,13 +849,18 @@ export class Dispatcher {
...
-        const existingRepository = matchExistingRepository(
+        let existingRepository = matchExistingRepository(
           state.repositories,
           path
         )
 
+        // in case this is valid git repository, there is no need to ask
+        // user for confirmation and it can be added automatically
+        if (!existingRepository && (await isGitRepository(path))) {
+          existingRepository = (await this.addRepositories([path]))[0]
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: l

File: app/src/lib/dispatcher/dispatcher.ts [hunk 6]
```diff
@@ -848,13 +849,18 @@ export class Dispatcher {
...
         const state = this.appStore.getState()
-        const existingRepository = matchExistingRepository(
+        let existingRepository = matchExistingRepository(
           state.repositories,
           path
         )
 
+        // in case this is valid git repository, there is no need to ask
+        // user for confirmation and it can be added automatically
+        if (!existingRepository && (await isGitRepository(path))) {
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: t

File: app/src/lib/dispatcher/dispatcher.ts [hunk 7]
```diff
@@ -848,13 +849,18 @@ export class Dispatcher {
...
+          existingRepository = (await this.addRepositories([path]))[0]
+        }
+
         if (existingRepository) {
```
Comment (truongdung3011) [inline app/src/lib/dispatcher/dispatcher.ts]: p

File: app/src/lib/dispatcher/dispatcher.ts (hunk 6 above)
Comment (shiftkey) [inline app/src/lib/dispatcher/dispatcher.ts]: :art: to be a bit more explicit: `existingRepository` -> `existingRepository == null`

Comment (shiftkey) [reply app/src/lib/dispatcher/dispatcher.ts]: I'm also fine with making this loop a bit more explicit, for better readability. Here's what I settled on for the flow:

```ts
if (existingRepository == null) {
  const isRepository = await isGitRepository(path)
  if (isRepository) {
    const addedRepositories = await this.addRepositories([path])
    existingRepository = addedRepositories[0]
  }
}
```

Comment (iljasergijenko) [reply app/src/lib/dispatcher/dispatcher.ts]: environment

Comment (iljasergijenko) [reply app/src/lib/dispatcher/dispatcher.ts]: environment

Comment (ggajos) [reply app/src/lib/dispatcher/dispatcher.ts]: Thanks @shiftkey, I've applied your idea.

**Comment by ggajos** [comment]
@shiftkey I believe it's ready to go

**Comment by shiftkey** [comment]
Great stuff @ggajos, thanks for the contribution!

**Comment by ggajos** [comment]
@shiftkey thanks :)

**Pull request merged**